publications/radstack/radstack.tex

```diff
@@ -0,0 +1,1148 @@
+over a year is better partitioned by day, and a data set with timestamps spread
+over a day is better partitioned by hour. 
+
+Segments are uniquely identified by a data source identifier, the time interval
+of the data, and a version string that increases whenever a new segment is
+created. The version string indicates the freshness of segment data; segments
+with later versions have newer views of data (over some time range) than
+segments with older versions. This segment metadata is used by the system for
+concurrency control; read operations always access data in a particular time
+range from the segments with the latest version identifiers for that time
+range.
+
+Druid segments are stored in a column orientation. Given that Druid is best
+used for aggregating event streams, the advantages of storing aggregate
+information as columns rather than rows are well documented
+\cite{abadi2008column}. Column storage allows for more efficient CPU usage as
+only what is needed is actually loaded and scanned.  In a row oriented data
+store, all columns associated with a row must be scanned as part of an
+aggregation. The additional scan time can introduce significant performance
+degradations \cite{abadi2008column}.
+
+Druid nodes use one thread to scan one segment at a time, and the amount of
+data that can be scanned in parallel is directly correlated to the number of
+available cores in the cluster. Segments are immutable, and immutability
+confers a few advantages. Immutable segments enable read consistency, and
+multiple threads can scan the same segment at the same. This helps enable
+higher read throughput. 
+
+A single query may scan thousands of segments concurrently, and
+many queries may run at the same time. We want to ensure that the entire
+cluster is not starved out while a single expensive query is executing. Thus,
+segments have an upper limit in how much data they can hold, and are sized
+to be scanned in a few milliseconds. By keeping segment computation very fast,
+cores and other resources are constantly being yielded. This ensures segments
+from different queries are always being scanned.
+
+Druid segments are very self-contained for the time interval of data that they
+hold. Column data is stored directly in the segment. Druid has multiple column
+types to represent various data formats. Timestamps are stored in long columns,
+dimensions are stored in string columns, and metrics are stored in int, float,
+long or double columns. Depending on the column type, different compression
+methods may be used. Metric columns are compressed using
+LZ4\cite{collet2013lz4} compression. String columns are dictionary encoded,
+similar to other data stores such as PowerDrill\cite{hall2012processing}.
+Additional indexes may be created for particular columns. For example, Druid
+will by default create inverted indexes for string columns.
+
+\subsection{Exploratory Analytics}
```

Comment:
This sub section seems to be called "Exploratory Analytics" but the initial paragraphs make me think that it is about inverted indexes.